import type { AuthenticatedRequest } from "@middlewares/auth";
import { logger } from "@services/logger";
import { isErrorWithMessage } from "@utils/error";
import type { Response } from "express";

import { CART_ERRORS } from "./cart.constant";
import {
  addItemService,
  deleteItemService,
  getItemsByUserIdService,
  getSummaryByUserIdService,
  updateCartService,
} from "./cart.service";

export const addItemAction = async (
  req: AuthenticatedRequest,
  res: Response
): Promise<void> => {
  logger.info("üõí Adding item to cart:", {
    userId: req.userId,
    body: req.body,
  });

  try {
    const cartItem = await addItemService(req.userId!, req.body);
    logger.info("‚úÖ Item added to cart:", cartItem);
    res.status(201).json(cartItem);
  } catch (error) {
    logger.error("‚ùå Failed to add item to cart:", {
      message: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
    });

    if (isErrorWithMessage(error, CART_ERRORS.PRODUCT_NOT_FOUND)) {
      res.status(404).json({ error: CART_ERRORS.PRODUCT_NOT_FOUND });
      return;
    }

    res.status(500).json({ error: "Internal Server Error" });
  }
};

export const updateItemAction = async (
  req: AuthenticatedRequest,
  res: Response
): Promise<void> => {
  logger.info("üìù Updating cart item:", {
    userId: req.userId,
    itemCartId: req.body.productId,
    update: req.body,
  });

  try {
    const updatedItem = await updateCartService(
      req.userId!,
      req.body.productId,
      req.body.quantity
    );

    logger.info("‚úÖ Cart item updated:", updatedItem);
    res.json(updatedItem);
  } catch (error) {
    logger.error("‚ùå Failed to update cart item:", {
      message: error instanceof Error ? error.message : String(error),
    });

    if (isErrorWithMessage(error, CART_ERRORS.ITEM_NOT_FOUND)) {
      res.status(404).json({ error: CART_ERRORS.ITEM_NOT_FOUND });
      return;
    }

    res.status(500).json({ error: "Internal Server Error" });
    return;
  }
};

export const deleteItemAction = async (
  req: AuthenticatedRequest,
  res: Response
): Promise<void> => {
  logger.info("üóëÔ∏è Removing cart item:", {
    userId: req.userId,
    itemCartId: req.params.itemCartId,
  });

  try {
    await deleteItemService(req.userId!, req.params.itemCartId);

    logger.info("‚úÖ Cart item deleted:", { itemCartId: req.params.itemCartId });
    res.status(204).send();
  } catch (error) {
    logger.error("‚ùå Failed to delete cart item:", {
      message: error instanceof Error ? error.message : String(error),
    });

    if (isErrorWithMessage(error, CART_ERRORS.ITEM_NOT_FOUND)) {
      res.status(404).json({ error: CART_ERRORS.ITEM_NOT_FOUND });
      return;
    }

    res.status(500).json({ error: "Internal Server Error" });
  }
};

export const getSummaryByUserIdAction = async (
  req: AuthenticatedRequest,
  res: Response
): Promise<void> => {
  logger.info("üì¶ Fetching cart summary for user:", {
    userId: req.params.userId,
  });

  try {
    const cartItems = await getSummaryByUserIdService(req.params.userId);
    logger.info("‚úÖ Cart items retrieved:", cartItems);
    res.json(cartItems);
  } catch (error) {
    logger.error("‚ùå Failed to fetch cart items:", {
      message: error instanceof Error ? error.message : String(error),
    });

    res.status(500).json({ error: "Internal Server Error" });
  }
};

export const getItemsByUserIdAction = async (
  req: AuthenticatedRequest,
  res: Response
): Promise<void> => {
  logger.info("üì¶ Fetching cart items for user:", {
    userId: req.params.userId,
  });

  try {
    const cartItems = await getItemsByUserIdService(req.params.userId);
    logger.info("‚úÖ Cart items retrieved:", cartItems);
    res.json(cartItems);
  } catch (error) {
    logger.error("‚ùå Failed to fetch cart items:", {
      message: error instanceof Error ? error.message : String(error),
    });

    res.status(500).json({ error: "Internal Server Error" });
  }
};
